### ISSUE TITLE
'broccoli' is not recognized as an internal or external command

### ISSUE THREAD

**Issue by justingosan** [tjhart/broccoli-couchdb-versioning#1]
Hi!

First of all, I'd like to thank you for this project as I've been looking for a solution such as this for a while now. Also, node is pretty new to me so forgive me if this is a non-issue.

I've got it all setup: installed the package through npm and created 'Brocfile.js' containing my local couchdb credentials. As mentioned in the README doc, I tried running `brocolli build` and `brocolli serve` but I get this error. I tried running `node brocfile.js` but it gives me this error `SyntaxError: Invalid regular expression: /\((map)|(reduce))$/: Unmatched ')'`.

I'm on Windows 8.1. I'll continue to explore the code but do you have any ideas on this in the meantime?

**Comment by justingosan** [comment]
I found the 2 causes.
1. `process.env.TMPDIR` isn't set by default (at least in Windows). My quick workaround now is to put the temp directory directly into `self.srcDir`
2. `var JS_PATH = new RegExp(path.sep + '((map)|(reduce))$');` directly translated to `\((map)|(reduce))$` in Windows which caused the regex error. My quick fix for now is to remove `path.sep +`. This seems to work fine. Our of curiousity, what was the purpose of prefixing JS_PATH with path.sep?

Also if in case anyone's wondering, I ran the code using `node brocfile.js` after quick fixing these said issues.

Thanks again for open-sourcing this project!

**Comment by justingosan** [comment]
I got `initDesign` working as it should but I cannot seem to push changes from the file system to couchdb. I see `CouchDBVersioning.prototype.read` is responsible for pushing the changes but I cannot seem to find it being called anywhere. Am I missing something?

**Comment by tjhart** [comment]
Justin,

Thanks for the feedback. A few thoughts:

`broccoli-couchdb-versioning` is a [broccoli](https://github.com/broccolijs/broccoli) plugin. It's Broccoli that reads `Brocfile.js` and runs the plugin when it notices that files have changed. Check out the link for more information. Once you have broccoli installed, you can run broccoli in 2 ways:

`broccoli serve` - this will keep broccoli running, and `broccoli-couchdb-versioning` will update CouchDB any time a file under your couchdb directory changes.

`broccoli build <build dir>` - This is a 'one and done' scenario. The directory is read, any changes it detects will be pushed, and then broccoli will quit.

`TMPDIR` not being defined - That's good information. I haven't done Windows in a while. I thought I remembered TMPDIR being defined in the shell environment by default. I'll do some research and see what I can come up with. 

`JS_PATH` - I want to write the `map` and `reduce` keys in the design document as javascript files. But keys like `foomap` or `bareduce` still need to be text files. That's why the path separator needs to be in the front of the regular expression. I need to escape the backslash since it's a special regular expression character.

I'll keep this issue open until I can find a resolution for both of those issues.

If you have trouble with `broccoli`, feel free to contact their team. They're good people.

**Comment by justingosan** [comment]
Oh man.. all the while I thought `broccoli` is the name of this utility! Thank you for clarifying extensively. Will get back to you when I've setup everything properly this time around.

Thank you again for the work you've put on this wonderful (and much needed) tool.

**Comment by tjhart** [comment]
@justingosan 

I have a few fixes in place, but I don't have a Windows environment in which to test. Would you mind temporarily changing your `package.json` entry for `broccoli-couchdb-versioning` to

"broccoli-couchdb-versioning":"https://github.com/tjhart/broccoli-couchdb-versioning#6789393"

and reporting on the results? It should now use the `TMP` environment variable in Windows for the temporary directory location, and the Windows path separator should be escaped for regular expressions.

**Comment by justingosan** [comment]
@tjhart 

I just setup everything now and the main functions work fine (retrieving design and saving from file system to couchdb). I am running into some issues though:
1. The map and reduce functions save as `.txt` files
   The `JS_PATH` regex string now looks like this `\\\(\(map\)\|\(reduce\)\)\$` which doesn't seem right. For now, I reverted to the previous commit without the `path.sep` which works fine with how I laid out my _design docs.
2. The function of manageDocs fails completely (although I do not make use of this functionality).
   Here are the logs:

```
CouchDB Versioning: Rebuilding records views
CouchDB Versioning: Rebuilding status views
CouchDB Versioning: Rebuilding validations views
Trace: CouchDB Versioning: ERROR: Sat Dec 13 2014 11:47:29 GMT+0800 (Malay Penin
sula Standard Time) [TypeError: Object.keys called on non-object]
    at D:\Projects\Node\moclad\node_modules\broccoli-couchdb-versioning\index.js
:319:15
    at $$$internal$$tryCatch (D:\Projects\Node\moclad\node_modules\broccoli-couc
hdb-versioning\node_modules\rsvp\dist\rsvp.js:470:16)
    at $$$internal$$invokeCallback (D:\Projects\Node\moclad\node_modules\broccol
i-couchdb-versioning\node_modules\rsvp\dist\rsvp.js:482:17)
    at $$$internal$$publish (D:\Projects\Node\moclad\node_modules\broccoli-couch
db-versioning\node_modules\rsvp\dist\rsvp.js:453:11)
    at $$$internal$$publishRejection (D:\Projects\Node\moclad\node_modules\brocc
oli-couchdb-versioning\node_modules\rsvp\dist\rsvp.js:395:7)
    at $$rsvp$asap$$flush (D:\Projects\Node\moclad\node_modules\broccoli-couchdb
-versioning\node_modules\rsvp\dist\rsvp.js:1531:9)
    at process._tickCallback (node.js:419:13)
```

I will try to debug these issues on my free time.

Thanks!

**Comment by justingosan** [comment]
In addition to the 2nd one above:

I had to create `couchdb\docs` folder as it doesn't seem to create on its own like the others (`couchdb\`, `couchdb\_design`, etc). Logs:

```
Warning: failed to stat couchdb\docs
Trace: CouchDB Versioning: ERROR: Sat Dec 13 2014 11:55:45 GMT+0800 (Malay Penin
sula Standard Time) { [Error: ENOENT, readdir 'D:\Projects\Node\moclad\couchdb\d
ocs']
  errno: 34,
  code: 'ENOENT',
```

**Comment by tjhart** [comment]
@justingosan 

Ooops. Yeah - that wasn't intended. I'll update the code to only escape the path separator.

 `couchdb/_design` is only created if you set 'initDesign' to true. But you've identified an interesting issue. Users may choose to use this tool to _either_ manage docs _or_ design docs, not necessarily both.

I should be able to get this fixed by the first of the week.

**Comment by tjhart** [comment]
@justingosan 

Hey Justin,

Sorry for the delay - I've been dealing with Carpel Tunnel, so it's been tough to type.

I've got another solution checked in at `#bf2cfba`. Would you mind testing it again?

**Comment by justingosan** [comment]
@tjhart 

Hi Tim, I apologize for the delay in my reply. Happy to inform you that this `1. The map and reduce functions save as .txt files` has been fixed! Thank you!

..but this still exists `2. The function of manageDocs fails completely`.. throwing the same error as before.

I've dealt with CTS before and it's always better to get it managed before it gets any worse :) Take it easy and Happy Holidays!

**Comment by tjhart** [comment]
@justingosan 

First, Happy New Year! With the holidays behind us, I'm happy to get back into resolving this issue.

I think the `2.` is, unfortunately, a communication/documentation issue. 

The `manageDocs` functionality, unlike `initDesign`, will never pull documents from a CouchDB into the directory structure. It will only ensure that existing documents in a `docs` subdirectory will be synchronized with the database. `broccoli-couchdb-versioning` will not create a `docs` subdirectory if it does not exist. If `docs` does exist, and `manageDocs` is set to true, then the system assumes there is at least one document in the directory that needs to be synchronized with the database.

The reason is that static documents often are an extremely small subset of all the data in a couchDB database. There's no practical way of determining which documents are interesting, and which documents are user data documents. Also, CouchDB has a very robust replication mechanism for synchronizing production data between databases.

Since you have no static documents to manage, you should not have a `docs` subdirectory, and of course, manageDocs should either not be declared, set to false.

Example:

``` javascript
//Brocfile.js
var couchVersioner = require('broccoli-couchdb-versioning);

module.exports = couchVersioner('couchdb', {
  url: 'http://localhost:5984/myDB',
  username: 'me',
  password: 'mine',
  initDesign:false
});
```

And your directory structure should look like this:

```
couchdb
+---- _design
     +---- myDesignDocDir
```

Of course, if I'm missing something, please let me know. In the mean time, I'll see if I can make this information more explicit in the README.

**Comment by justingosan** [comment]
Hi @tjhart 

Happy New Year to you as well! Thank you for clarifying that. I delved into the code and #2's error wasn't after all related to the functionality of `manageDocs`. I have a design doc which I only use for validations and only contains the `validate_doc_update` key (no `views` key) which I found causes the error. Also, I've found that even though the error is logged, everything works as they should :)
